### PR TITLE
GFORMS-1418 Output summary & acknowledgement pdfs using Gform pdf gen…

### PR DESCRIPTION
**Pull request by girishkamat**
…erator

### CODE REVIEW

File: project/Dependencies.scala [hunk 1]
```diff
@@ -29,7 +29,8 @@ object Dependencies {
     "com.nrinaudo"                %% "kantan.csv"                 % "0.5.1",
     "com.miguelfonseca.completely" % "completely-core"            % "0.8.0",
     "org.jsoup"                    % "jsoup"                      % "1.13.1",
-    "org.webjars.npm"              % "govuk-frontend"             % "3.11.0"
+    "org.webjars.npm"              % "govuk-frontend"             % "3.11.0",
+    "com.openhtmltopdf"           % "openhtmltopdf-pdfbox"      % "1.0.10"
```
Comment (josef-vlach) [inline project/Dependencies.scala:33]: Please run `scalafmtSbt` to fix formatting

File: test/uk/gov/hmrc/gform/models/FormModelSupport.scala [hunk 2]
```diff
@@ -100,13 +100,13 @@ trait FormModelSupport extends GraphSpec {
...
-  def mkFormModelOptics(
+  def mkFormModelOptics[D <: DataOrigin](
     formTemplate: FormTemplate,
     data: VariadicFormData[SourceOrigin.OutOfDate]
-  )(implicit messages: Messages, lang: LangADT): FormModelOptics[DataOrigin.Browser] = {
+  )(implicit messages: Messages, lang: LangADT): FormModelOptics[D] = {
     val authCache: AuthCacheWithForm = mkAuthCacheWithForm(formTemplate)
     FormModelOptics
-      .mkFormModelOptics[DataOrigin.Browser, Id, SectionSelectorType.Normal](data, authCache, recalculation)
+      .mkFormModelOptics[D, Id, SectionSelectorType.Normal](data, authCache, recalculation)
```
Comment (josef-vlach) [inline test/uk/gov/hmrc/gform/models/FormModelSupport.scala:109]: Can we revert this changes in `mkFormModelOptics`. Going abstract means loosing useful information, and I don't think is good price to pay for test sake.

Use `isInstanceOf` in the test code if needed.

File: app/uk/gov/hmrc/gform/gformbackend/GformBackEndService.scala [hunk 3]
```diff
@@ -217,13 +226,21 @@ class GformBackEndService(
...
+          formModelOpticsUpdated,
+          cache.formTemplate.destinations match {
+            case DestinationList(_, acknowledgementSection, _) =>
+              acknowledgementSection.instructionPdf.map(p => PDFModel.HeaderFooter(p.header, p.footer))
+            case _ => None
+          },
+          submissionDetails
         )
-        .map(Some(_))
+        .map(Option(_))
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/gformbackend/GformBackEndService.scala:243]: Why was this changed to `Option.apply` ?

File: app/uk/gov/hmrc/gform/pdf/model/PDFModel.scala [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    new PDFCustomRender[PDFType.Summary] {
+
+      override val layout: PDFLayout = PDFLayout.Default
+
+      override def getPageTitle(page: Page[_], maybePresentationHint: Option[PresentationHint])(implicit
+        lise: SmartStringEvaluator
+      ): Option[String] =
+        maybePresentationHint
+          .filter(_ == InvisiblePageTitle)
+          .fold[Option[String]](page.shortName.orElse(Option(page.title)).map(_.value()))(_ => None)
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/pdf/model/PDFModel.scala:96]: `Some` should be enough here

File: app/uk/gov/hmrc/gform/pdf/model/PDFModel.scala [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+      override def getFormComponentLabel(formComponent: FormComponent)(implicit
+        lise: SmartStringEvaluator
+      ): Option[String] =
+        formComponent.shortName.map(_.value()).orElse(Option(formComponent.label.value()))
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/pdf/model/PDFModel.scala:101]: `Some` should be enough here

File: app/uk/gov/hmrc/gform/pdf/model/PDFPageModelBuilder.scala [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+      val maybePresentationHint = source.fold { _ =>
+        singleton.page.presentationHint
+      } { _ =>
+        singleton.page.presentationHint
+      } { addToList =>
+        addToList.presentationHint
+          .orElse(singleton.page.presentationHint)
+      }
+      val pageTitle = getPageTitle(singleton.page, maybePresentationHint)
+      Option(PageData(pageTitle, pageFields, sectionNumber.value.toString))
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/pdf/model/PDFPageModelBuilder.scala:154]: Where this habit is coming from. In this particular case param to `Option.apply` will never be `null` so why don't use `Some` instead?

File: app/uk/gov/hmrc/gform/gform/SummaryController.scala [hunk 7]
```diff
@@ -216,12 +221,22 @@ class SummaryController(
...
     ) { implicit request => implicit l => cache => implicit sse => formModelOptics =>
-      pdfService
-        .generateSummaryPDF(maybeAccessCode, cache, SummaryPagePurpose.ForUser, formModelOptics)
-        .map { pdfStream =>
+      pdfRenderService
+        .createPDFHtml[DataOrigin.Mongo, SectionSelectorType.Normal, PDFType.Summary](
+          request.messages.messages(
+            "summary.checkYourAnswers"
+          ) + " - " + cache.formTemplate.formName.value + " - GOV.UK",
+          Option(cache.formTemplate.summarySection.title.value()),
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/gform/SummaryController.scala:229]: And here

File: app/uk/gov/hmrc/gform/gform/SummaryController.scala [hunk 8]
```diff
@@ -216,12 +221,22 @@ class SummaryController(
...
+          cache,
+          formModelOptics,
+          Option(PDFModel.HeaderFooter(Option(cache.formTemplate.summarySection.header), None)),
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/gform/SummaryController.scala:232]: And here
And here

File: app/uk/gov/hmrc/gform/gform/SummaryController.scala [hunk 9]
```diff
@@ -216,12 +221,22 @@ class SummaryController(
...
+          None
+        )
+        .flatMap(pdfGeneratorService.generatePDFLocal)
+        .map { pdfBytes =>
           Result(
             header = ResponseHeader(200, Map.empty),
-            body = HttpEntity.Streamed(pdfStream, None, Some("application/pdf"))
+            body = HttpEntity.Strict(ByteString(pdfBytes), Some("application/pdf"))
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/gform/SummaryController.scala:239]: Just a question, why streaming abandoned?

**Review by josef-vlach — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/gform/gform/SummaryController.scala (hunk 9 above)
Comment (girishkamat) [reply app/uk/gov/hmrc/gform/gform/SummaryController.scala:239]: Was struggling a bit to get Stream working with java.io. API, I think I have a working piece of code now.

File: app/uk/gov/hmrc/gform/gform/SummaryController.scala (hunk 8 above)
Comment (girishkamat) [reply app/uk/gov/hmrc/gform/gform/SummaryController.scala:232]: Done

File: app/uk/gov/hmrc/gform/gform/SummaryController.scala (hunk 7 above)
Comment (girishkamat) [reply app/uk/gov/hmrc/gform/gform/SummaryController.scala:229]: Done

File: app/uk/gov/hmrc/gform/pdf/model/PDFPageModelBuilder.scala (hunk 6 above)
Comment (girishkamat) [reply app/uk/gov/hmrc/gform/pdf/model/PDFPageModelBuilder.scala:154]: My lack of knowledge on the diff between Option(x) and Some(x).

File: app/uk/gov/hmrc/gform/pdf/model/PDFModel.scala (hunk 5 above)
Comment (girishkamat) [reply app/uk/gov/hmrc/gform/pdf/model/PDFModel.scala:101]: Yes

File: app/uk/gov/hmrc/gform/pdf/model/PDFModel.scala (hunk 4 above)
Comment (girishkamat) [reply app/uk/gov/hmrc/gform/pdf/model/PDFModel.scala:96]: Yes

File: app/uk/gov/hmrc/gform/gformbackend/GformBackEndService.scala (hunk 3 above)
Comment (girishkamat) [reply app/uk/gov/hmrc/gform/gformbackend/GformBackEndService.scala:243]: Reverted

File: test/uk/gov/hmrc/gform/models/FormModelSupport.scala (hunk 2 above)
Comment (girishkamat) [reply test/uk/gov/hmrc/gform/models/FormModelSupport.scala:109]: Makes sense, corrected

File: project/Dependencies.scala (hunk 1 above)
Comment (girishkamat) [reply project/Dependencies.scala:33]: Done

**Review by josef-vlach — APPROVED** [review]
